// @textlint/kernel/lib/src/descriptor/TextlintFilterRuleDescriptor.js
// LICENSE : MIT
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextlintFilterRuleDescriptor = void 0;
const rule_creator_helper_1 = require("./rule-creator-helper");
const fast_equals_1 = require("fast-equals");
/**
 * Textlint Rule Descriptor.
 * It handles RuleCreator and RuleOption.
 */
class TextlintFilterRuleDescriptor {
    constructor(kernelFilterRule) {
        this.kernelFilterRule = kernelFilterRule;
    }
    get id() {
        return this.kernelFilterRule.ruleId;
    }
    get rule() {
        return this.kernelFilterRule.rule;
    }
    /**
     * Return true if this rule is enabled.
     */
    get enabled() {
        return this.rawOptions !== false;
    }
    /**
     * Return filter function
     * You should check hasLiner before call this.
     */
    get filter() {
        return (0, rule_creator_helper_1.getFilter)(this.rule);
    }
    /**
     * Return normalized rule option object.
     * If the rule have not options, return `true` by default.
     */
    get normalizedOptions() {
        // default: { ruleName: true }
        const DefaultRuleConfigValue = {};
        if (typeof this.kernelFilterRule.options === "boolean" || this.kernelFilterRule.options === undefined) {
            return DefaultRuleConfigValue;
        }
        else {
            return this.kernelFilterRule.options;
        }
    }
    get rawOptions() {
        return this.kernelFilterRule.options;
    }
    /**
     * Return true if descriptor is same
     */
    equals(descriptor) {
        return this.rule === descriptor.rule && (0, fast_equals_1.deepEqual)(this.normalizedOptions, descriptor.normalizedOptions);
    }
    toKernel() {
        return this.kernelFilterRule;
    }
    toJSON() {
        return {
            id: this.id,
            options: this.normalizedOptions
        };
    }
}
exports.TextlintFilterRuleDescriptor = TextlintFilterRuleDescriptor;
