// textlint-rule-sentence-length/lib/test/sentence-length-test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var textlint_tester_1 = __importDefault(require("textlint-tester"));
var sentence_length_1 = __importDefault(require("../src/sentence-length"));
var tester = new textlint_tester_1.default();
tester.run("textlint-rule-sentence-length", sentence_length_1.default, {
    valid: [
        "This is a article",
        "Test`code`です。",
        {
            text: '"This" is code.',
            options: {
                max: 15
            }
        },
        // Exception: A link in the Paragraph
        "[textlint/textlint-filter-rule-comments: textlint filter rule that disables all rules between comments directive.](https://github.com/textlint/textlint-filter-rule-comments)",
        "- [textlint/textlint-filter-rule-comments: textlint filter rule that disables all rules between comments directive.](https://github.com/textlint/textlint-filter-rule-comments)",
        {
            text: "> ignore",
            options: {
                max: 1
            }
        },
        {
            // == 12345
            text: "ab`c`de",
            options: {
                max: 5
            }
        },
        {
            // == 12345
            text: '[123](http://example.com "123456")45',
            options: {
                max: 5
            }
        },
        {
            // ignore /()$/。
            text: "1234(56789)",
            options: {
                max: 5,
                skipPatterns: ["/\\(.*\\)$/"]
            }
        },
        {
            // html node
            // == 12345
            text: "<s>123</s><b>45</b>",
            options: {
                max: 5
            }
        },
        {
            // url string link
            text: "[http://example.com](http://example.com)",
            options: {
                max: 5
            }
        },
        {
            // List
            text: '- [abc](http://example.com "abc")de',
            options: {
                max: 5
            }
        },
        {
            // List
            text: "\n- abcde\n- abcde\n- abcde\n- abcde\n- abcde\n",
            options: {
                max: 5
            }
        },
        {
            text: "\u5B9F\u884C\u30B3\u30F3\u30C6\u30AD\u30B9\u30C8\u304C\"Script\"\u3067\u3042\u308B\u5834\u5408\u3001\u305D\u306E\u30B3\u30FC\u30C9\u76F4\u4E0B\u306B\u66F8\u304B\u308C\u305F`this`\u306F\u30B0\u30ED\u30FC\u30D0\u30EB\u30AA\u30D6\u30B8\u30A7\u30AF\u30C8\u3092\u53C2\u7167\u3057\u307E\u3059\u3002\n\u30B0\u30ED\u30FC\u30D0\u30EB\u30AA\u30D6\u30B8\u30A7\u30AF\u30C8\u3068\u306F\u3001\u5B9F\u884C\u74B0\u5883\u306B\u304A\u3044\u3066\u7570\u306A\u308B\u3082\u306E\u304C\u5B9A\u7FA9\u3055\u308C\u3066\u3044\u307E\u3059\u3002\n\u30D6\u30E9\u30A6\u30B6\u306A\u3089`window`\u30AA\u30D6\u30B8\u30A7\u30AF\u30C8\u3001Node.js\u306A\u3089`global`\u30AA\u30D6\u30B8\u30A7\u30AF\u30C8\u3068\u306A\u308A\u307E\u3059\u3002",
            options: {
                max: 90
            }
        },
        {
            text: "\u307E\u305F*CORS policy Invalid*\u306E\u3088\u3046\u306A\u30A8\u30E9\u30FC\u304C\u30B3\u30F3\u30BD\u30FC\u30EB\u306B\u8868\u793A\u3055\u308C\u3066\u3044\u308B\u5834\u5408\u306F\u3001[Same Origin Policy][]\u306B\u3088\u308A`index.js`\u306E\u8AAD\u307F\u8FBC\u307F\u304C\u5931\u6557\u3057\u3066\u3044\u307E\u3059\u3002\n\n[Same Origin Policy]: https://example.com/",
            options: {
                max: 90
            }
        },
        // regression test
        // https://github.com/textlint-rule/textlint-rule-sentence-length/issues/13
        {
            text: "ではみなさんは、そういうふうに川だと云いわれたり、乳の流れたあとだと云われたりしていたこのぼんやりと白いものがほんとうは何かご承知ですか。\n" +
                "\n" +
                'ではみなさんは、そういうふうに川だと云いわれたり、乳の流れたあとだと云われたりしていたこの<a href="https://www.aozora.gr.jp/cards/000081/files/456_15050.html" target="_blank" rel="noopener noreferrer">ぼんやりと白いもの</a>がほんとうは何かご承知ですか。\n'
        },
        {
            // Issue: https://github.com/textlint-rule/textlint-rule-sentence-length/issues/38
            text: "JavaScript\u3092Wasm\u306B\u3057\u3066\u5B9F\u884C\u3059\u308B\u4ED5\u7D44\u307F\u306B\u3064\u3044\u3066\u3002\nQuickJS\u3092\u4F7F\u3044JavaScript\u3092ByteCode\u306B\u3057\u305Fxxxx\u3092\u4F5C\u6210\u3057\u3001`QuickJS.wasm`\u3068\u52D5\u7684\u306B\u30EA\u30F3\u30AF\u3057\u3066\u5927\u90E8\u5206\u3092\u5171\u6709\u3057\u3066\u3044\u308B\u3002\nShopify Function\u3067\u5229\u7528\u3055\u308C\u308B\u304C\u3001\u975E\u540C\u671F\u51E6\u7406\u306E\u5236\u9650\u30845ms\u672A\u6E80\u3067\u306E\u5B9F\u884C\u5236\u9650\u304C\u3042\u308B\u3002"
        },
        // https://github.com/textlint-rule/textlint-rule-sentence-length/issues/18
        {
            // length: 35
            // skip "..." -> `He said .`
            text: "He said \"This is a pen. I like it\".",
            options: {
                max: 10,
                skipPatterns: ['/".*"/']
            }
        },
        {
            text: "𦥑井と臼井",
            options: {
                max: 5,
                countBy: "codepoints"
            }
        }
    ],
    invalid: [
        {
            text: "123456",
            options: {
                max: 5
            },
            errors: [
                {
                    message: "Line 1 sentence length(6) exceeds the maximum sentence length of 5.\nOver 1 characters."
                }
            ]
        },
        {
            text: "123\n45",
            options: {
                max: 5
            },
            errors: [
                {
                    message: "Line 1 sentence length(6) exceeds the maximum sentence length of 5.\nOver 1 characters."
                }
            ]
        },
        {
            text: "text\n\n123456",
            options: {
                max: 5
            },
            errors: [
                {
                    message: "Line 3 sentence length(6) exceeds the maximum sentence length of 5.\nOver 1 characters."
                }
            ]
        },
        {
            // test: https://github.com/azu/textlint-rule-sentence-length/issues/5
            text: "11111\n2222\n3333333",
            options: {
                max: 5
            },
            errors: [
                {
                    message: "Line 1 sentence length(18) exceeds the maximum sentence length of 5.\nOver 13 characters."
                }
            ]
        },
        {
            // test: https://github.com/azu/textlint-rule-sentence-length/issues/5
            text: "\u3042\u3044\u3046\u3048\u304A\u3002\n\n\u3042\u3044\u3046\u3048\u304A\u3002\n\u3042\u3044\u3046\u3048\u304A\u3042\u304B\u3055\u305F\u306A\u3002",
            options: {
                max: 10
            },
            errors: [
                {
                    message: "Line 4 sentence length(11) exceeds the maximum sentence length of 10.\nOver 1 characters.",
                    line: 4,
                    column: 1
                }
            ]
        },
        {
            text: "text\n\nline3\nline4\n",
            options: {
                max: 5
            },
            errors: [
                {
                    message: "Line 3 sentence length(11) exceeds the maximum sentence length of 5.\nOver 6 characters."
                }
            ]
        },
        // example
        {
            text: "This is text.\n            \n_Middleware_ \u3068\u3044\u3046\u4ED5\u7D44\u307F\u81EA\u4F53\u306F[Connect](../connect/README.md)\u3068\u4F3C\u3066\u3044\u307E\u3059\u3002\n\u3057\u304B\u3057\u3001 _Middleware_ \u304C\u76F4\u63A5\u7684\u306B\u7D50\u679C(State)\u3092\u76F4\u63A5\u66F8\u304D\u63DB\u3048\u308B\u4E8B\u306F\u3067\u304D\u307E\u305B\u3093\u3002\n\u3053\u308C\u306F\u3001Connect\u304C\u6700\u7D42\u7684\u306A\u7D50\u679C(response)\u3092\u66F8\u304D\u63DB\u3048\u3067\u304D\u308B\u306E\u5BFE\u3057\u3066\u3001\nRedux\u306E _Middleware_ \u306F\u6271\u3048\u308B\u7BC4\u56F2\u304Cdispatch\u304B\u3089Reducer\u307E\u3067\u3068\u7DDA\u5F15\u3055\u308C\u3066\u3044\u308B\u9055\u3044\u3068\u8A00\u3048\u307E\u3059\u3002",
            options: {
                max: 100
            },
            errors: [
                {
                    message: "Line 5 sentence length(102) exceeds the maximum sentence length of 100.\nOver 2 characters.",
                    line: 5,
                    column: 1
                }
            ]
        },
        {
            // ignore /()$/。
            text: "123456789(56789)",
            options: {
                max: 5,
                skipPatterns: ["/\\(.*\\)$/"]
            },
            errors: [
                {
                    message: "Line 1 sentence length(9, original:16) exceeds the maximum sentence length of 5.\n" +
                        "Over 4 characters."
                }
            ]
        },
        {
            // url string link
            text: "TEST [http://example.com](http://example.com)",
            errors: [
                {
                    message: "Line 1 sentence length(23) exceeds the maximum sentence length of 5.\nOver 18 characters."
                }
            ],
            options: {
                max: 5,
                skipUrlStringLink: false
            }
        },
        {
            text: "𦥑井と臼井",
            errors: [
                {
                    message: "Line 1 sentence length(6) exceeds the maximum sentence length of 5.\nOver 1 characters."
                }
            ],
            options: {
                max: 5
            }
        }
    ]
});
