// @textlint/kernel/lib/src/descriptor/TextlintPluginDescriptors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextlintPluginDescriptors = void 0;
const DescriptorUtil_js_1 = require("./DescriptorUtil.js");
/**
 * Collection class of TextlintPluginDescriptor
 */
class TextlintPluginDescriptors {
    constructor(pluginDescriptorList = []) {
        this.pluginDescriptorList = pluginDescriptorList;
    }
    /**
     * Return enabled descriptors
     */
    get descriptors() {
        return this.pluginDescriptorList.filter((descriptor) => {
            return descriptor.enabled;
        });
    }
    /**
     * Return all descriptors that include disabled descriptors
     */
    get allDescriptors() {
        return this.pluginDescriptorList;
    }
    /**
     * Return available extensions
     * It is not includes disabled plugin's extensions.
     */
    get availableExtensions() {
        return this.descriptors.reduce((extensions, descriptor) => {
            return extensions.concat(descriptor.availableExtensions);
        }, []);
    }
    /**
     * find PluginDescriptor with extension.
     * This is forward match.
     * @param {string} extension - file extension (ex. ".md", ".txt")
     */
    findPluginDescriptorWithExt(extension) {
        return this.descriptors.find((descriptor) => {
            return descriptor.availableExtensions.some((availableExtension) => extension.endsWith(availableExtension));
        });
    }
    /**
     * filter duplicated descriptors
     */
    withoutDuplicated() {
        const newDescriptorList = (0, DescriptorUtil_js_1.filterDuplicateDescriptor)(this.pluginDescriptorList);
        return new TextlintPluginDescriptors(newDescriptorList);
    }
    /**
     * Convert this to TextlintKernel rules format
     * @returns {Array}
     */
    toKernelPluginsFormat() {
        return this.descriptors.map((descriptor) => {
            return descriptor.toKernel();
        });
    }
    toJSON() {
        return this.descriptors.map((descriptor) => {
            return descriptor.toJSON();
        });
    }
}
exports.TextlintPluginDescriptors = TextlintPluginDescriptors;
