// @textlint/config-loader/lib/src/config-loader.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadRawConfig = exports.loadConfig = exports.loadPackagesFromRawConfig = void 0;
const rc_config_loader_1 = require("rc-config-loader");
const textlint_module_resolver_js_1 = require("./textlint-module-resolver.js");
const loader_js_1 = require("./loader.js");
const node_buffer_1 = require("node:buffer");
const node_fs_1 = __importDefault(require("node:fs"));
/**
 * Load packages in RawConfig and return loaded config object
 * @param options
 */
const loadPackagesFromRawConfig = async (options) => {
    var _a, _b, _c, _d, _e, _f;
    // TODO: validation
    // Search textlint's module
    const moduleResolver = new textlint_module_resolver_js_1.TextLintModuleResolver({
        rulesBaseDirectory: options.node_moduleDir
    });
    // rules
    const { rules, rulesError } = await (0, loader_js_1.loadRules)({
        rulesObject: (_a = options.rawConfig.rules) !== null && _a !== void 0 ? _a : {},
        moduleResolver,
        testReplaceDefinitions: (_b = options.testReplaceDefinitions) === null || _b === void 0 ? void 0 : _b.rule
    });
    // filterRules
    const { filterRules, filterRulesError } = await (0, loader_js_1.loadFilterRules)({
        rulesObject: (_c = options.rawConfig.filters) !== null && _c !== void 0 ? _c : {},
        moduleResolver,
        testReplaceDefinitions: (_d = options.testReplaceDefinitions) === null || _d === void 0 ? void 0 : _d.filterRules
    });
    // plugins
    const { plugins, pluginsError } = await (0, loader_js_1.loadPlugins)({
        pluginsObject: (_e = options.rawConfig.plugins) !== null && _e !== void 0 ? _e : {},
        moduleResolver,
        testReplaceDefinitions: (_f = options.testReplaceDefinitions) === null || _f === void 0 ? void 0 : _f.plugins
    });
    if (rulesError) {
        return {
            ok: false,
            error: rulesError
        };
    }
    if (filterRulesError) {
        return {
            ok: false,
            error: filterRulesError
        };
    }
    if (pluginsError) {
        return {
            ok: false,
            error: pluginsError
        };
    }
    const loadedConfig = {
        rules,
        plugins,
        filterRules
    };
    // TODO: after validation
    return {
        ok: true,
        config: loadedConfig
    };
};
exports.loadPackagesFromRawConfig = loadPackagesFromRawConfig;
/**
 *  Load config file and return config object that is loaded rule instance.
 * @param options
 */
const loadConfig = async (options) => {
    const rawResult = await (0, exports.loadRawConfig)(options);
    if (!rawResult.ok) {
        return {
            ok: false,
            error: rawResult.error
        };
    }
    const packageOptions = {
        rawConfig: rawResult.rawConfig,
        node_moduleDir: options.node_modulesDir,
        testReplaceDefinitions: options.testReplaceDefinitions
    };
    const result = await (0, exports.loadPackagesFromRawConfig)(options.preLoadingPackage ? options.preLoadingPackage(packageOptions) : packageOptions);
    if (!result.ok) {
        return {
            ok: false,
            configFilePath: rawResult.configFilePath,
            error: result.error
        };
    }
    return {
        ok: true,
        config: result.config,
        configFilePath: rawResult.configFilePath,
        rawConfig: rawResult.rawConfig
    };
};
exports.loadConfig = loadConfig;
/**
 *  Load config file and return parsed config object that is not loaded rule instance
 *  It is just JSON present for config file. Raw data
 * @param options
 */
const loadRawConfig = async (options) => {
    try {
        const results = (0, rc_config_loader_1.rcFile)("textlint", {
            cwd: options.cwd,
            configFileName: options.configFilePath,
            packageJSON: {
                fieldName: "textlint"
            }
        });
        // Not Found
        if (!results) {
            return {
                ok: false,
                error: {
                    message: "textlint config is not found",
                    errors: [
                        new Error(`textlint config is not found

textlint require .textlintrc config file.
The config file define the use of rules.`)
                    ]
                }
            };
        }
        if (results.filePath && node_fs_1.default.existsSync(results.filePath) && !(0, node_buffer_1.isUtf8)(node_fs_1.default.readFileSync(results.filePath))) {
            throw new Error("textlint configuration file must be encoded in UTF-8");
        }
        return {
            ok: true,
            rawConfig: results.config,
            configFilePath: results.filePath
        };
    }
    catch (error) {
        return {
            ok: false,
            error: {
                message: "textlint config is not found",
                errors: [error instanceof Error ? error : new Error(String(error))]
            }
        };
    }
};
exports.loadRawConfig = loadRawConfig;
