// @textlint/config-loader/lib/src/is.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTextlintRulePresetCreator = exports.isTextlintFilterRuleModule = exports.isTextlintRuleModule = void 0;
exports.hasLinter = hasLinter;
exports.hasFixer = hasFixer;
// Type guards
function isObjectWithProperty(obj, property) {
    return typeof obj === "object" && obj !== null && property in obj;
}
/**
 * detect that ruleCreator has linter function
 * @param {*} ruleCreator
 * @returns {boolean}
 */
function hasLinter(ruleCreator) {
    if (!ruleCreator) {
        return false;
    }
    if (isObjectWithProperty(ruleCreator, "linter") && typeof ruleCreator.linter === "function") {
        return true;
    }
    if (typeof ruleCreator === "function") {
        return true;
    }
    return false;
}
/**
 * detect that ruleCreator has fixer function
 * @param {*} ruleCreator
 * @returns {boolean}
 */
function hasFixer(ruleCreator) {
    if (!ruleCreator) {
        return false;
    }
    return (isObjectWithProperty(ruleCreator, "fixer") && typeof ruleCreator.fixer === "function" && hasLinter(ruleCreator));
}
/**
 * is textlint rule module
 * type guard
 * @param mod
 */
const isTextlintRuleModule = (mod) => {
    // it is same with isRuleModule() in @textlint/kernel
    return hasLinter(mod) || hasFixer(mod);
};
exports.isTextlintRuleModule = isTextlintRuleModule;
/**
 * is textlint filter rule reporter
 * type guard
 * @param mod
 */
const isTextlintFilterRuleModule = (mod) => {
    return typeof mod === "function";
};
exports.isTextlintFilterRuleModule = isTextlintFilterRuleModule;
const hasOwnProperty = Object.prototype.hasOwnProperty;
/**
 * is textlint rule preset
 * type guard
 * @param mod
 */
const isTextlintRulePresetCreator = (mod) => {
    return hasOwnProperty.call(mod, "rules") && hasOwnProperty.call(mod, "rulesConfig");
};
exports.isTextlintRulePresetCreator = isTextlintRulePresetCreator;
