// @textlint/config-loader/lib/src/loader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadRules = exports.loadFilterRules = exports.loadPlugins = void 0;
exports.loadPreset = loadPreset;
const config_util_js_1 = require("./config-util.js");
const module_interop_1 = require("@textlint/module-interop");
const is_js_1 = require("./is.js");
const utils_1 = require("@textlint/utils");
const resolver_1 = require("@textlint/resolver");
const isPluginCreator = (mod) => {
    return typeof mod === "object" && mod !== null && Object.prototype.hasOwnProperty.call(mod, "Processor");
};
const loadPlugins = async ({ pluginsObject, moduleResolver, testReplaceDefinitions }) => {
    const plugins = [];
    const pluginErrors = [];
    if (Array.isArray(pluginsObject)) {
        // { plugins: ["a", "b"] }
        await Promise.all(pluginsObject.map(async (pluginId) => {
            var _a;
            const resolvedModule = moduleResolver.resolvePluginPackageName(pluginId);
            const mod = await (0, resolver_1.dynamicImport)(resolvedModule.filePath, {
                parentModule: "config-loader"
            });
            const plugin = (0, module_interop_1.moduleInterop)((_a = mod.exports) === null || _a === void 0 ? void 0 : _a.default);
            if (!isPluginCreator(plugin)) {
                pluginErrors.push(new Error(`Plugin should be object that has "Processor" property. But "${pluginId}" is not.

Please check "${pluginId}" is valid plugin.
FilePath: ${resolvedModule.filePath}

For more details, See FAQ: https://github.com/textlint/textlint/blob/master/docs/faq/failed-to-load-textlints-module.md
`));
                return;
            }
            plugins.push({
                type: "Plugin",
                pluginId,
                plugin,
                filePath: resolvedModule.filePath,
                moduleName: resolvedModule.moduleName,
                inputModuleName: resolvedModule.inputModuleName
            });
        }));
    }
    else {
        // { plugins: { "a": true, "b": options } }
        await Promise.all(Object.entries(pluginsObject).map(async ([pluginId, pluginOptions]) => {
            var _a;
            try {
                // Test Replace logic
                const replacedDefinition = testReplaceDefinitions &&
                    testReplaceDefinitions.find((definition) => {
                        return definition.pluginId === pluginId;
                    });
                if (replacedDefinition) {
                    // for debug
                    plugins.push(replacedDefinition);
                }
                else {
                    const resolvedPlugin = moduleResolver.resolvePluginPackageName(pluginId);
                    const mod = await (0, resolver_1.dynamicImport)(resolvedPlugin.filePath, {
                        parentModule: "config-loader"
                    });
                    const plugin = (0, module_interop_1.moduleInterop)((_a = mod.exports) === null || _a === void 0 ? void 0 : _a.default);
                    if (!isPluginCreator(plugin)) {
                        pluginErrors.push(new Error(`Plugin should be object that has "Processor" property. But "${pluginId}" is not.

Please check "${pluginId}" is valid plugin.
FilePath: ${resolvedPlugin.filePath}

For more details, See FAQ: https://github.com/textlint/textlint/blob/master/docs/faq/failed-to-load-textlints-module.md
`));
                        return;
                    }
                    plugins.push({
                        type: "Plugin",
                        pluginId,
                        plugin,
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        options: pluginOptions,
                        filePath: resolvedPlugin.filePath,
                        moduleName: resolvedPlugin.moduleName,
                        inputModuleName: resolvedPlugin.inputModuleName
                    });
                }
            }
            catch (error) {
                pluginErrors.push(error);
            }
        }));
    }
    return {
        plugins,
        pluginsError: pluginErrors.length === 0
            ? null
            : {
                message: "Can not load plugin",
                errors: pluginErrors
            }
    };
};
exports.loadPlugins = loadPlugins;
const loadFilterRules = async ({ rulesObject, moduleResolver, testReplaceDefinitions }) => {
    // rules
    const rules = [];
    const ruleErrors = [];
    await Promise.all(Object.entries(rulesObject).map(async ([ruleId, ruleOptions]) => {
        var _a;
        try {
            // Test Replace logic
            const replacedDefinition = testReplaceDefinitions &&
                testReplaceDefinitions.find((definition) => {
                    return definition.ruleId === ruleId;
                });
            if (replacedDefinition) {
                // for debug
                rules.push(replacedDefinition);
            }
            else {
                const resolvePackage = moduleResolver.resolveFilterRulePackageName(ruleId);
                const mod = await (0, resolver_1.dynamicImport)(resolvePackage.filePath, {
                    parentModule: "config-loader"
                });
                const ruleModule = (0, module_interop_1.moduleInterop)((_a = mod.exports) === null || _a === void 0 ? void 0 : _a.default);
                if (!(0, is_js_1.isTextlintFilterRuleModule)(ruleModule)) {
                    ruleErrors.push(new Error(`Filter rule should be object that has "filter" property. But ${ruleId} is not.`));
                    return;
                }
                // rule
                rules.push({
                    ruleId,
                    rule: ruleModule,
                    options: ruleOptions,
                    filePath: resolvePackage.filePath,
                    moduleName: resolvePackage.moduleName,
                    inputModuleName: resolvePackage.inputModuleName
                });
            }
        }
        catch (error) {
            ruleErrors.push(error);
        }
    }));
    return {
        filterRules: rules,
        filterRulesError: ruleErrors.length === 0
            ? null
            : {
                message: "Can not load filter rule",
                errors: ruleErrors
            }
    };
};
exports.loadFilterRules = loadFilterRules;
const loadRules = async ({ rulesObject, moduleResolver, testReplaceDefinitions }) => {
    // rules
    const rules = [];
    const ruleErrors = [];
    await Promise.all(Object.entries(rulesObject).map(async ([ruleId, ruleOptions]) => {
        var _a;
        try {
            // Test Replace logic
            const replacedDefinition = testReplaceDefinitions &&
                testReplaceDefinitions.find((definition) => {
                    return definition.ruleId === ruleId;
                });
            // if rule is disabled, skip to load
            if (!ruleOptions) {
                return;
            }
            if (replacedDefinition) {
                // for debug
                rules.push(replacedDefinition);
            }
            else {
                if ((0, config_util_js_1.isPresetRuleKey)(ruleId)) {
                    // load preset
                    const presetRulesOptions = typeof ruleOptions === "boolean" ? {} : ruleOptions;
                    const rulesInPreset = await loadPreset({
                        presetName: ruleId,
                        presetRulesOptions,
                        moduleResolver
                    });
                    rules.push(...rulesInPreset);
                }
                else {
                    // load rule
                    const resolvePackage = moduleResolver.resolveRulePackageName(ruleId);
                    const mod = await (0, resolver_1.dynamicImport)(resolvePackage.filePath, {
                        parentModule: "config-loader"
                    });
                    const ruleModule = (0, module_interop_1.moduleInterop)((_a = mod.exports) === null || _a === void 0 ? void 0 : _a.default);
                    if (!(0, is_js_1.isTextlintRuleModule)(ruleModule)) {
                        ruleErrors.push(new Error(`Rule should have "rules" and "rulesConfig" properties. But ${ruleId} is not.
                        
Please check ${ruleId} is valid rule.
FilePath: ${resolvePackage.filePath}

For more details, See FAQ: https://github.com/textlint/textlint/blob/master/docs/faq/failed-to-load-textlints-module.md                        
`));
                        return;
                    }
                    // rule
                    rules.push({
                        type: "Rule",
                        ruleId,
                        rule: ruleModule,
                        options: ruleOptions,
                        filePath: resolvePackage.filePath,
                        moduleName: resolvePackage.moduleName,
                        inputModuleName: resolvePackage.inputModuleName
                    });
                }
            }
        }
        catch (error) {
            ruleErrors.push(error);
        }
    }));
    return {
        rules,
        rulesError: ruleErrors.length === 0
            ? null
            : {
                message: "Can not load rule",
                errors: ruleErrors
            }
    };
};
exports.loadRules = loadRules;
/**
 * prefer .textlintrc config than preset.rulesConfig
 * Do not merge configs - user config takes complete precedence when provided
 *
 * Rule otions priority:
 * 1. If userRuleConfig is defined, use user config
 * 2. If userRuleConfig is false, use user config - the rule is disabled
 * 3. If presetRuleConfig is true or presetRulesOptions[ruleKey] is not false, use preset config
 *    - If presetRulesOptions[ruleKey] is false by default, user can enable the rule by setting it to true
 **/
const getRuleOptions = ({ userRuleConfig, presetRulesConfig }) => {
    // If user config is false, the rule is disabled
    if (userRuleConfig === false) {
        return false;
    }
    // If user config is defined, use user config
    if (userRuleConfig !== true && userRuleConfig !== undefined) {
        return userRuleConfig;
    }
    // If user config is true, use preset config
    if (userRuleConfig === true) {
        return presetRulesConfig;
    }
    // In other cases, use preset config
    return presetRulesConfig;
};
async function loadPreset({ presetName, presetRulesOptions, moduleResolver }) {
    var _a;
    const presetPackageName = moduleResolver.resolvePresetPackageName(presetName);
    const mod = await (0, resolver_1.dynamicImport)(presetPackageName.filePath, {
        parentModule: "config-loader"
    });
    const preset = (0, module_interop_1.moduleInterop)((_a = mod.exports) === null || _a === void 0 ? void 0 : _a.default);
    if (!(0, is_js_1.isTextlintRulePresetCreator)(preset)) {
        throw new Error(`preset should have rules and rulesConfig: ${presetName}`);
    }
    // we should use preset.rules → some preset use different name actual rule
    /*
      // "textlint-rule-preset-example/index.js"
      {
        "rules": {

          "a: ruleA,
        },
        "rulesConfig": {
          "a": true
        }
      }

      - type: "RuleInPreset"
      - ruleId: "preset-example/rule-a"
      - options: true
      - filePath: path to "textlint-rule-preset-example/index.js"
      - moduleName: "textlint-rule-preset-example"
      - ruleKey: "a",
      - inputModuleName: "preset-example"
     */
    return Object.keys(preset.rules).map((ruleKey) => {
        var _a, _b;
        const normalizedKey = (0, utils_1.normalizeTextlintPresetSubRuleKey)({ preset: presetName, rule: ruleKey });
        return {
            type: "RuleInPreset",
            ruleId: normalizedKey,
            rule: preset.rules[ruleKey],
            options: getRuleOptions({
                userRuleConfig: (_a = presetRulesOptions[ruleKey]) !== null && _a !== void 0 ? _a : true,
                presetRulesConfig: (_b = preset.rulesConfig[ruleKey]) !== null && _b !== void 0 ? _b : true
            }),
            filePath: presetPackageName.filePath,
            // preset package name
            moduleName: presetPackageName.moduleName,
            inputModuleName: presetPackageName.inputModuleName,
            // rule key in preset
            ruleKey
        };
    });
}
