// @textlint/config-loader/lib/src/textlint-module-resolver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextLintModuleResolver = void 0;
const path = __importStar(require("node:path"));
const package_prefix_js_1 = require("./package-prefix.js");
const textlint_package_name_util_js_1 = require("./textlint-package-name-util.js");
const resolver_1 = require("@textlint/resolver");
/**
 * This class aim to resolve textlint's package name and get the module path.
 *
 * Define
 *
 * - `package` is npm package
 * - `module` is package's main module
 *
 * ## Support
 *
 * - textlint-rule-*
 * - textlint-preset-*
 * - textlint-plugin-*
 * - textlint-config-*
 */
class TextLintModuleResolver {
    constructor(config) {
        this.moduleCache = new Map();
        this.tryResolvePackagePath = (modulePath) => {
            const cachedFilePath = this.moduleCache.get(modulePath);
            if (cachedFilePath) {
                return cachedFilePath;
            }
            const ret = (0, resolver_1.tryResolve)(modulePath, {
                parentModule: "config-loader"
            });
            if (ret) {
                this.moduleCache.set(modulePath, ret);
                return ret;
            }
            return null;
        };
        /**
         * @type {string} baseDirectory for resolving
         */
        this.baseDirectory = config && config.rulesBaseDirectory ? config.rulesBaseDirectory : "";
    }
    /**
     * Take package name, and return path to module.
     * @param {string} packageName
     * @returns {string} return path to module
     */
    resolveRulePackageName(packageName) {
        const baseDir = this.baseDirectory;
        const fullPackageName = (0, textlint_package_name_util_js_1.createFullPackageName)(package_prefix_js_1.PackageNamePrefix.rule, packageName);
        // <rule-name> or textlint-rule-<rule-name>
        const resultFullPackagePath = this.tryResolvePackagePath(path.join(baseDir, fullPackageName));
        if (resultFullPackagePath) {
            return {
                inputModuleName: packageName,
                moduleName: fullPackageName,
                filePath: resultFullPackagePath
            };
        }
        const resultPackagePath = this.tryResolvePackagePath(path.join(baseDir, packageName));
        if (resultPackagePath) {
            return {
                inputModuleName: packageName,
                moduleName: packageName,
                filePath: resultPackagePath
            };
        }
        throw new ReferenceError(`Failed to load textlint's rule module: "${packageName}" is not found.
See FAQ: https://github.com/textlint/textlint/blob/master/docs/faq/failed-to-load-textlints-module.md
`);
    }
    /**
     * Take package name, and return path to module.
     * @param {string} packageName
     * @returns {string} return path to module
     */
    resolveFilterRulePackageName(packageName) {
        const baseDir = this.baseDirectory;
        const fullPackageName = (0, textlint_package_name_util_js_1.createFullPackageName)(package_prefix_js_1.PackageNamePrefix.filterRule, packageName);
        // <rule-name> or textlint-filter-rule-<rule-name> or @scope/<rule-name>
        const resultFullPackagePath = this.tryResolvePackagePath(path.join(baseDir, fullPackageName));
        if (resultFullPackagePath) {
            return {
                inputModuleName: packageName,
                moduleName: fullPackageName,
                filePath: resultFullPackagePath
            };
        }
        const resultPackagePath = this.tryResolvePackagePath(path.join(baseDir, packageName));
        if (resultPackagePath) {
            return {
                inputModuleName: packageName,
                moduleName: packageName,
                filePath: resultPackagePath
            };
        }
        throw new ReferenceError(`Failed to load textlint's filter rule module: "${packageName}" is not found.
See FAQ: https://github.com/textlint/textlint/blob/master/docs/faq/failed-to-load-textlints-module.md
`);
    }
    /**
     * Take package name, and return path to module.
     * @param {string} packageName
     * @returns {string} return path to module
     */
    resolvePluginPackageName(packageName) {
        const baseDir = this.baseDirectory;
        const fullPackageName = (0, textlint_package_name_util_js_1.createFullPackageName)(package_prefix_js_1.PackageNamePrefix.plugin, packageName);
        // <plugin-name> or textlint-plugin-<rule-name>
        const resultFullPackagePath = this.tryResolvePackagePath(path.join(baseDir, fullPackageName));
        if (resultFullPackagePath) {
            return {
                inputModuleName: packageName,
                moduleName: fullPackageName,
                filePath: resultFullPackagePath
            };
        }
        const resultPackagePath = this.tryResolvePackagePath(path.join(baseDir, packageName));
        if (resultPackagePath) {
            return {
                inputModuleName: packageName,
                moduleName: packageName,
                filePath: resultPackagePath
            };
        }
        throw new ReferenceError(`Failed to load textlint's plugin module: "${packageName}" is not found.
See FAQ: https://github.com/textlint/textlint/blob/master/docs/faq/failed-to-load-textlints-module.md
`);
    }
    /**
     * Take package name, and return path to module.
     * @param {string} packageName
     * The user must specify preset- prefix to these `packageName`.
     */
    resolvePresetPackageName(packageName) {
        const baseDir = this.baseDirectory;
        const PREFIX = package_prefix_js_1.PackageNamePrefix.rulePreset;
        /* Implementation Note

        preset name is defined in config file:
        In the case, `packageName` is "preset-gizmo"
        TextLintModuleResolver resolve "preset-gizmo" to "textlint-rule-preset-gizmo"
        {
            "rules": {
                "preset-gizmo": {
                    "ruleA": false
                }
            }
        }
         */
        // preset-<name> or textlint-rule-preset-<name>
        // @scope/preset-<name> or @scope/textlint-rule-preset-<name>
        const packageNameWithoutPreset = packageName
            .replace(/^preset-/, "")
            .replace(/^@([^/]+)\/preset-(.*)$/, `@$1/$2`);
        const fullPackageName = (0, textlint_package_name_util_js_1.createFullPackageName)(PREFIX, packageNameWithoutPreset);
        const fullFullPackageName = `${PREFIX}${packageNameWithoutPreset}`;
        // textlint-rule-preset-<preset-name> or @scope/textlint-rule-preset-<preset-name>
        const resultFullPresetPackagePath = this.tryResolvePackagePath(path.join(baseDir, fullFullPackageName));
        if (resultFullPresetPackagePath) {
            return {
                inputModuleName: packageName,
                moduleName: fullFullPackageName,
                filePath: resultFullPresetPackagePath
            };
        }
        // <preset-name>
        const resultPresetPackagePath = this.tryResolvePackagePath(path.join(baseDir, packageNameWithoutPreset));
        if (resultPresetPackagePath) {
            return {
                inputModuleName: packageName,
                moduleName: packageNameWithoutPreset,
                filePath: resultPresetPackagePath
            };
        }
        // <rule-name>
        const resultFullPackagePath = this.tryResolvePackagePath(path.join(baseDir, fullPackageName));
        if (resultFullPackagePath) {
            return {
                inputModuleName: packageName,
                moduleName: fullPackageName,
                filePath: resultFullPackagePath
            };
        }
        // <package-name>
        const resultPackagePath = this.tryResolvePackagePath(path.join(baseDir, packageName));
        if (resultPackagePath) {
            return {
                inputModuleName: packageName,
                moduleName: packageName,
                filePath: resultPackagePath
            };
        }
        throw new ReferenceError(`Failed to load textlint's preset module: "${packageName}" is not found.
See FAQ: https://github.com/textlint/textlint/blob/master/docs/faq/failed-to-load-textlints-module.md
`);
    }
    /**
     * Take Config package name, and return path to module.
     * @param {string} packageName
     * @returns {string} return path to module
     */
    resolveConfigPackageName(packageName) {
        const baseDir = this.baseDirectory;
        const fullPackageName = (0, textlint_package_name_util_js_1.createFullPackageName)(package_prefix_js_1.PackageNamePrefix.config, packageName);
        // <plugin-name> or textlint-config-<rule-name>
        const pkgPath = (0, resolver_1.tryResolve)(path.join(baseDir, fullPackageName), {
            parentModule: "config-loader"
        }) ||
            (0, resolver_1.tryResolve)(path.join(baseDir, packageName), {
                parentModule: "config-loader"
            });
        if (!pkgPath) {
            throw new ReferenceError(`Failed to load textlint's config module: "${packageName}" is not found.
See FAQ: https://github.com/textlint/textlint/blob/master/docs/faq/failed-to-load-textlints-module.md
`);
        }
        return pkgPath;
    }
}
exports.TextLintModuleResolver = TextLintModuleResolver;
