// @textlint/kernel/lib/src/util/createDummyTextLintResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDummyTextLintResult = void 0;
// TODO: Fix can not support
/**
 * Create dummy TextLintResult
 * @param message
 * @param filePath
 */
const createDummyTextLintResult = (message, filePath) => {
    return {
        filePath: filePath !== null && filePath !== void 0 ? filePath : "<Unknown>",
        messages: [
            {
                message,
                type: "lint",
                loc: {
                    start: {
                        line: 1,
                        column: 0
                    },
                    end: {
                        line: 1,
                        column: 0
                    }
                },
                index: 0,
                line: 1,
                column: 0,
                range: [0, 1],
                severity: 2,
                ruleId: "plugin-error"
            }
        ]
    };
};
exports.createDummyTextLintResult = createDummyTextLintResult;
